File: #1043.partition-array-for-maximum-sum.cpp
class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k) {

        int size = arr.size();
        vector<int> result(size);
        for (int i = 0; i < size; i++) {
            int max = 0;
            for (int j = 1; j <= k && i - (j - 1) >= 0; j++) {
                max = std::max(max, arr[i-(j-1)]);
                result[i] = std::max(result[i],
                     ((i-j >= 0) ? result[i-j] : 0) + max * j);
            }
            cout << result[i] << endl;
        }
        return result[size - 1];
    }
};

File: #1043.partition-array-for-maximum-sum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */
var maxSumAfterPartitioning = function(arr, k) {
    const length = arr.length;
    let dp = new Array(length + 1).fill(0);
    for (let i = length - 1; i >= 0; i--) {
        let maxSum = -Infinity;
        let len = 0;
        let max = -Infinity;
        for (let j = i; j < Math.min(length, i + k); j++) {
            len++;
            max = Math.max(max, arr[j]);
            maxSum = Math.max(maxSum, ((len * max) + dp[j + 1]));
        }
        dp[i] = maxSum;
    }
    return dp[0];
};
